DDJ-200 support for 4 decks

fkbreitl opened the pull request.

fkbreitl committed:
support for 4 decks

Holzhaus commented:
Does this PR change anything from the user's perspective? If so, please document that with a PR to the manual repo.

fkbreitl commented:
Everything in the manual is still valid. There are some additional features which I will add to the manual soon. In the meantime the files could already be updated.Am 20.10.2020 17:34 schrieb Jan Holthuis <notifications@github.com>:
Does this PR change anything from the user's perspective? If so, please document that with a PR to the manual repo.

—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.

Holzhaus commented:
> In the meantime the files could already be updated.

Without the documentation, we can't review this properly. Also, I think the policy should be to only merge changes if the documentation is ready, too. Otherwise contributors might be busy or lose interested and we end up with even undocumented or not properly documented controller mappings in the mixxx repo :D

fkbreitl committed:
Improvements from CodeFactor

fkbreitl commented:
Ok, I have updated the documentation (the PR is still open) and fixed all issues from CodeFactor.
Code should be ready for review.

Holzhaus commented:
> I have updated the documentation (the PR is still open)

Can you post a link? I don't see the PR.

fkbreitl commented:
Sorry, it always goes to your repo. Here is the correct one: https://github.com/mixxxdj/manual/pull/237

Holzhaus commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -1,41 +1,126 @@
+    DDJ200.LEDsOff();
+
+    // start with focus on library for selecting tracks (delay seems required)
+    engine.beginTimer(500, "engine.setValue(\"[Library]\", " +
+                      "\"MoveFocus\", 1);", true);
```
Please don't use a string here, pass a function expression.

Holzhaus commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -1,41 +1,126 @@
+DDJ200.init = function() {
+    // creating associative arrays for the virtual decks
+    this.vDeck = {
+        1: this.vDeck1, 2: _.clone(this.vDeck1),
+        3: _.clone(this.vDeck1), 4: _.clone(this.vDeck1)
```
I'm really not a fan of lodash. This is very unreadable IMHO.

fkbreitl replied:
What alternative do you suggest?

Holzhaus replied:
Something like this maybe?

```javascript
this.vDeck = {}
for (var i = 1; i <= 4; i++) {
    this.vDeck[i] = {
        syncEnabled: false,
        volMSB: 0,
        rateMSB: 0,
        jogDisabled: false,
    }
}
```

Holzhaus commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -1,41 +1,126 @@
+    for (var i = 1; i <= 4; i++) {
+        var vgroup = "[Channel" + i + "]";
+        // run onTrackLoad after every track load to e.g. set LEDs accordingly
+        engine.connectControl(vgroup, "track_loaded", "DDJ200.onTrackLoad");
```
Please use `engine.makeConnection` instead and don't use a string, pass the actual function.

Holzhaus commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -1,41 +1,126 @@
+        var deckNo = script.deckFromGroup(group);
+        var vDeckNo = DDJ200.vDeckNo[deckNo];
+        var vgroup = "[Channel" + vDeckNo +"]";
+        engine.setValue(vgroup, "LoadSelectedTrack", true);
```
Use `script.triggerControl` here.

Holzhaus commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -1,41 +1,126 @@
-        engine.beginTimer(900, "DDJ200.jogDisabled["+deckNumber+"] = false;",
-            true);
+        // enable jog after 900 ms again
+        engine.beginTimer(900, "DDJ200.vDeck[" + vDeckNo +
+                          "][\"jogDisabled\"] = false;", true);
```
use a function expression

Holzhaus commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -46,14 +131,219 @@ DDJ200.seek = function(channel, control, value, status, group) {
+            if (deckNo === 1) {
+                // toggle virtual deck of controller deck 1
+                DDJ200.vDeckNo[1] = 4 - DDJ200.vDeckNo[1];
+                if (DDJ200.vDeckNo[1] === 1) LED = 0;
```
missing braces

Holzhaus commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -46,14 +131,219 @@ DDJ200.seek = function(channel, control, value, status, group) {
+            } else { // deckNo === 2
+                // toggle virtual deck of controller deck 2
+                DDJ200.vDeckNo[2] = 6 - DDJ200.vDeckNo[2];
+                if (DDJ200.vDeckNo[2] === 2) LED = 0;
```
missing braces

Holzhaus commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -46,14 +131,219 @@ DDJ200.seek = function(channel, control, value, status, group) {
+            0x7F * engine.getValue(vgroup, "pfl"));
+    }
+
+    if (vDeckNo % 2) c = 7; else c = 9;
```
missing braces

Holzhaus reviewed (changes requested):
Thank you. Some first comments.

Holzhaus commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -46,14 +131,219 @@ DDJ200.seek = function(channel, control, value, status, group) {
+        var vgroup = "[Channel" + vDeckNo +"]";
+        var pfl = ! engine.getValue(vgroup, "pfl");
+        engine.setValue(vgroup, "pfl", pfl);
+        if (DDJ200.fourDeckMode === false) {
```
```suggestion
        if (!DDJ200.fourDeckMode) {
```

Holzhaus commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -46,14 +131,219 @@ DDJ200.seek = function(channel, control, value, status, group) {
+
+DDJ200.switchLEDs = function(vDeckNo) {
+    // set LEDs of controller deck according to virtual deck
+    var c = 1; if (vDeckNo % 2) c = 0;
```
missing braces and multiple statements on the same line. Also, please use more meaningful variable names, e.g.: `channelOffset` or something like that.

fkbreitl committed:
Changes suggested by Holzhaus

fkbreitl reviewed (commented):
I have don't the changes you requested

fkbreitl committed:
Changes from CodeFactor

fkbreitl committed:
More changes from CodeFactor

Holzhaus commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -1,41 +1,133 @@
+    // set LEDs (hotcues, etc.) for the loaded deck
+    // if controller is switched to this deck
+    var vDeckNo = script.deckFromGroup(vgroup);
+    var deckNo = 2; if (vDeckNo % 2) deckNo = 1;
```
missing braces and linebreak.

why not this?
```suggestion
    var deckNo = (vDeckNo % 2) ? 1 : 2;
```

Holzhaus commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -1,41 +1,133 @@
+};
+
+DDJ200.browseTracks = function(value) {
+    DDJ200.jogCounter += value-64;
```
```suggestion
    DDJ200.jogCounter += value - 64;
```

Holzhaus commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -1,41 +1,133 @@
     // Convert value down to +1/-1
     // Register the movement
-    engine.scratchTick(script.deckFromGroup(group), value - 64);
+    if (DDJ200.shiftPressed["left"] === true) { DDJ200.browseTracks(value); } else {
```
```suggestion
    if (DDJ200.shiftPressed["left"]) {
        DDJ200.browseTracks(value);
    } else {
```

Holzhaus commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -1,41 +1,133 @@
-    engine.scratchTick(script.deckFromGroup(group), value - 64);
+    if (DDJ200.shiftPressed["left"] === true) { DDJ200.browseTracks(value); } else {
+        var vDeckNo = DDJ200.vDeckNo[script.deckFromGroup(group)];
+        if (! DDJ200.vDeck[vDeckNo]["jogDisabled"]) {
```
```suggestion
        if (!DDJ200.vDeck[vDeckNo]["jogDisabled"]) {
```

Holzhaus commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -1,41 +1,133 @@
+                                  
+       // create associative arrays for 4 virtual decks
+        this.vDeck[i] = {
+        syncEnabled: false,
+        volMSB: 0,
+        rateMSB: 0,
+        jogDisabled: false,
```
It's a bit confusing that this is inverted.

How about this?
```suggestion
        syncEnabled: false,
        volMSB: 0,
        rateMSB: 0,
        jogEnabled: true,
```

Holzhaus commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -46,14 +138,219 @@ DDJ200.seek = function(channel, control, value, status, group) {
 };
+
+DDJ200.toggleFourDeckMode = function(channel, control, value) {
+    if (value) { // only if button pressed, not releases, i.e. value === 0
```
Everytime you only want to wrap the whole function body in `if (value) { ... }`, you could just early exit to reduce the nesting depth.

```javascript
if (!value) {
    return;
}
```

fkbreitl replied:
I know.

Holzhaus commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -46,14 +138,219 @@ DDJ200.seek = function(channel, control, value, status, group) {
+
+DDJ200.toggleFourDeckMode = function(channel, control, value) {
+    if (value) { // only if button pressed, not releases, i.e. value === 0
+        DDJ200.fourDeckMode = ! DDJ200.fourDeckMode;
```
```suggestion
        DDJ200.fourDeckMode = !DDJ200.fourDeckMode;
```

Holzhaus commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -46,14 +138,219 @@ DDJ200.seek = function(channel, control, value, status, group) {
+
+DDJ200.eq = function(channel, control, value, status, group) {
+    var val = script.absoluteNonLin(value, 0, 1, 4);
+    var eq = 1; if (control === 0x0B) eq = 2;
```
```suggestion
    var eq = (control === 0x0B) ? 2 : 1;
```

Holzhaus commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -46,14 +138,219 @@ DDJ200.seek = function(channel, control, value, status, group) {
+
+DDJ200.switchLEDs = function(vDeckNo) {
+    // set LEDs of controller deck according to virtual deck
+    var d = 1;  // d = deckNo (0 for left, 1 for right deck)
+    if (vDeckNo % 2) { d = 0; }
```
```suggestion
    // 0 for left, 1 for right deck
    var deckNo = (vDeckNo % 2) ? 0 : 1;
```

Holzhaus commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -46,14 +138,219 @@ DDJ200.seek = function(channel, control, value, status, group) {
+            DDJ200.LoadSelectedTrack(channel, control, value, status, group);
+        } else if (DDJ200.fourDeckMode) { //right shift + pfl 1/2 toggles
+            var deckNo = script.deckFromGroup(group);
+            var vDeckNo = 0;
```
```suggestion
            var vDeckNo;
```

Holzhaus commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -46,14 +138,219 @@ DDJ200.seek = function(channel, control, value, status, group) {
+        } else if (DDJ200.fourDeckMode) { //right shift + pfl 1/2 toggles
+            var deckNo = script.deckFromGroup(group);
+            var vDeckNo = 0;
+            var LED = 0x7F;
+            if (deckNo === 1) {
+                // toggle virtual deck of controller deck 1
+                DDJ200.vDeckNo[1] = 4 - DDJ200.vDeckNo[1];
+                if (DDJ200.vDeckNo[1] === 1) { LED = 0; }
+                vDeckNo = DDJ200.vDeckNo[1];
+            } else { // deckNo === 2
+                // toggle virtual deck of controller deck 2
+                DDJ200.vDeckNo[2] = 6 - DDJ200.vDeckNo[2];
+                if (DDJ200.vDeckNo[2] === 2) { LED = 0; }
+                vDeckNo = DDJ200.vDeckNo[2];
+            }
+            midi.sendShortMsg(status, 0x54, LED); // toggle virtual deck LED
```
```suggestion
            var led = 0x7F;
            if (deckNo === 1) {
                // toggle virtual deck of controller deck 1
                DDJ200.vDeckNo[1] = 4 - DDJ200.vDeckNo[1];
                if (DDJ200.vDeckNo[1] === 1) {
                    led = 0;
                }
                vDeckNo = DDJ200.vDeckNo[1];
            } else { // deckNo === 2
                // toggle virtual deck of controller deck 2
                DDJ200.vDeckNo[2] = 6 - DDJ200.vDeckNo[2];
                if (DDJ200.vDeckNo[2] === 2) {
                    led = 0;
                }
                vDeckNo = DDJ200.vDeckNo[2];
            }
            midi.sendShortMsg(status, 0x54, led); // toggle virtual deck LED
```

fkbreitl committed:
Update res/controllers/Pioneer-DDJ-200-scripts.js

Co-authored-by: Jan Holthuis <holthuis.jan@googlemail.com>

fkbreitl committed:
Update res/controllers/Pioneer-DDJ-200-scripts.js

Co-authored-by: Jan Holthuis <holthuis.jan@googlemail.com>

fkbreitl committed:
Update res/controllers/Pioneer-DDJ-200-scripts.js

Co-authored-by: Jan Holthuis <holthuis.jan@googlemail.com>

fkbreitl committed:
Update res/controllers/Pioneer-DDJ-200-scripts.js

Co-authored-by: Jan Holthuis <holthuis.jan@googlemail.com>

fkbreitl committed:
Update res/controllers/Pioneer-DDJ-200-scripts.js

Co-authored-by: Jan Holthuis <holthuis.jan@googlemail.com>

fkbreitl committed:
Update res/controllers/Pioneer-DDJ-200-scripts.js

Co-authored-by: Jan Holthuis <holthuis.jan@googlemail.com>

fkbreitl committed:
Update res/controllers/Pioneer-DDJ-200-scripts.js

Co-authored-by: Jan Holthuis <holthuis.jan@googlemail.com>

fkbreitl committed:
Update res/controllers/Pioneer-DDJ-200-scripts.js

Co-authored-by: Jan Holthuis <holthuis.jan@googlemail.com>

fkbreitl committed:
Update res/controllers/Pioneer-DDJ-200-scripts.js

Co-authored-by: Jan Holthuis <holthuis.jan@googlemail.com>

fkbreitl committed:
Update res/controllers/Pioneer-DDJ-200-scripts.js

Co-authored-by: Jan Holthuis <holthuis.jan@googlemail.com>

fkbreitl committed:
Merge branch '2.2' of https://github.com/mixxxdj/mixxx into ddj-200-4-deck

fkbreitl committed:
Merge branch 'ddj-200-4-deck' of https://github.com/fkbreitl/mixxx into ddj-200-4-deck

Holzhaus commented:
Note to myself: Use Squash & Merge strategy here as the commit messages don't really help.

fkbreitl committed:
Changes according to Holzhaus

fkbreitl committed:
Variable name fix

fkbreitl commented:
CodeFactor shows an error that does not exist in fkbreitl:ddj-200-4deck and hence that I cannot fix...

fkbreitl commented:
> 
> 
> CodeFactor shows an error that does not exist in fkbreitl:ddj-200-4deck and hence that I cannot fix...

Any ideas how to fix this?
I don't even understand how this code appears. It is not what I have committed and what should be there.

Holzhaus commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -1,41 +1,134 @@
+                                  
+       // create associative arrays for 4 virtual decks
+        this.vDeck[i] = {
+        syncEnabled: false,
+        volMSB: 0,
+        rateMSB: 0,
+        jogEnabled: true,
```
```suggestion
            syncEnabled: false,
            volMSB: 0,
            rateMSB: 0,
            jogEnabled: true,
```

Holzhaus commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -1,41 +1,134 @@
-DDJ200.init = function() {};
+DDJ200.init = function() {
+    for (var i = 1; i <= 4; i++) {
+                                  
```
```suggestion
```

Holzhaus commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -1,41 +1,134 @@
+    } else {
+        var vDeckNo = DDJ200.vDeckNo[script.deckFromGroup(group)];
+        if (DDJ200.vDeck[vDeckNo]["jogEnabled"]) {
+            var vgroup = "[Channel" + vDeckNo +"]";
```
```suggestion
            var vgroup = "[Channel" + vDeckNo + "]";
```

Holzhaus commented:
> > 
> > 
> > CodeFactor shows an error that does not exist in fkbreitl:ddj-200-4deck and hence that I cannot fix...
> 
> Any ideas how to fix this?
> I don't even understand how this code appears. It is not what I have committed and what should be there.

No idea, just ignore it.

fkbreitl committed:
Update res/controllers/Pioneer-DDJ-200-scripts.js

Co-authored-by: Jan Holthuis <holthuis.jan@googlemail.com>

fkbreitl commented:
Are we reviewing the right repository?

fkbreitl committed:
Update res/controllers/Pioneer-DDJ-200-scripts.js

Co-authored-by: Jan Holthuis <holthuis.jan@googlemail.com>

Holzhaus reviewed (changes requested):
Please apply this patch:

```diff
diff --git a/res/controllers/Pioneer-DDJ-200-scripts.js b/res/controllers/Pioneer-DDJ-200-scripts.js
index 91f8b4fd2c..50d61bd78b 100644
--- a/res/controllers/Pioneer-DDJ-200-scripts.js
+++ b/res/controllers/Pioneer-DDJ-200-scripts.js
@@ -8,8 +8,8 @@ var DDJ200 = {

 DDJ200.init = function() {
     for (var i = 1; i <= 4; i++) {
-
-       // create associative arrays for 4 virtual decks
+
+        // create associative arrays for 4 virtual decks
         this.vDeck[i] = {
             syncEnabled: false,
             volMSB: 0,
@@ -20,19 +20,19 @@ DDJ200.init = function() {

         // run onTrackLoad after every track load to set LEDs accordingly
         engine.makeConnection(vgroup, "track_loaded", function(ch, vgroup) {
-                DDJ200.onTrackLoad(ch, vgroup);
-            });
+            DDJ200.onTrackLoad(ch, vgroup);
+        });

         // set Pioneer CDJ cue mode for all decks
-         engine.setValue(vgroup, "cue_cdj", true);
+        engine.setValue(vgroup, "cue_cdj", true);
     }

     DDJ200.LEDsOff();

     // start with focus on library for selecting tracks (delay seems required)
     engine.beginTimer(500, function() {
-            engine.setValue("[Library]", "MoveFocus", 1);
-        }, true);
+        engine.setValue("[Library]", "MoveFocus", 1);
+    }, true);
 };

 DDJ200.shutdown = function() { DDJ200.LEDsOff(); };
@@ -125,8 +125,8 @@ DDJ200.touch = function(channel, control, value, status, group) {
     } else {
         // enable jog after 900 ms again
         engine.beginTimer(900, function() {
-                DDJ200.vDeck[vDeckNo]["jogEnabled"] = true;
-            }, true);
+            DDJ200.vDeck[vDeckNo]["jogEnabled"] = true;
+        }, true);
         // disable scratch
         engine.scratchDisable(vDeckNo);
     }
```

fkbreitl commented:
How can I apply it?

Holzhaus commented:
Save it to a file and use `git apply < /path/to/file`.

fkbreitl commented:
I guess I have to pull the changes from my git repo first and then copy and past the patch in the file (since I don't see a save option).

fkbreitl commented:
Unfortunately after the above:
`error: corrupt patch at line 6`

Holzhaus commented:
Are you sure you're in the correct branch?

Holzhaus commented:
Maybe it's a copy and paste issue. Here's a file you can download and use for `git apply`: [patch.txt](https://github.com/mixxxdj/mixxx/files/5433702/patch.txt)

fkbreitl committed:
Fixed Javascrip formatting issues

fkbreitl commented:
Maybe. Thank you.
I have now corrected my flymake indentation for Javascript and indented the whole buffer instead of applying your patch, which should have done the same thing.
And I fixed all remaining +"]".

fkbreitl commented:
I am also wondering what is the best way to keep my github repo up to date?
Should I pull the updates from mixxx to my laptop and then push them to my github repo or can I update github first and then pull it to my laptop, which would avoid the uploading part?

Holzhaus commented:
Why would you need to keep your GitHub repo up-to-date? I just have configured an `upstream` remote:

```
$ git remote add upstream https://github.com/mixxxdj/mixxx.git
```

When you start a new branch, just fetch the latest changes from the remote and start a new branch from there:

```
$ git fetch upstream
$ git checkout -b mynewbranch upstream/2.3
```

You only need to push your branch, there's no reason to update other branches:

```
$ git push origin mynewbranch
```

fkbreitl commented:
Aright. Thanks!

Holzhaus commented:
Unfortunately the indentation doesn't conform to our style rules yet, here's another patch: 
[patch.txt](https://github.com/mixxxdj/mixxx/files/5434581/patch.txt)

fkbreitl commented:
Oh no, what are your style rules called? Is it not Javascript?

fkbreitl committed:
applyed Mixxx indentation rule

fkbreitl commented:
How did you detect these issues? My pre-commit hook did not complain.

Holzhaus commented:
> How did you detect these issues? My pre-commit hook did not complain.

This branch is based on 2.2, but the pre-commit hooks only exist on 2.3 and onwards. I just merged this branch into 2.3 locally and ran the pre-commit hooks there, and used `git diff` to generate the patch.

fkbreitl commented:
Should I write a bug report that the Mixxx style rules do not comply with JS conventions?

Holzhaus commented:
> Should I write a bug report that the Mixxx style rules do not comply with JS conventions?

What do you mean? Because we don't use hanging indents?

fkbreitl commented:
Yes, I thought it was a rule, but apparently its not...

Holzhaus commented on res/controllers/Pioneer DDJ-200.midi.xml:
```diff
@@ -3,7 +3,7 @@
         <name>Pioneer DDJ-200</name>
         <author>Dan Giddins, Frank.Breitling at gmx.de</author>
-        <description>2-deck USB and Bluetooth MIDI controller.</description>
+        <description>2-deck USB and Bluetooth MIDI controller supporting 4 decks.</description>
```
Maybe this? Otherwise the sentence sounds like a paradox.

```suggestion
        <description>2-deck USB and Bluetooth MIDI controller with support for 4-deck mode.</description>
```

Holzhaus commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -46,14 +140,220 @@ DDJ200.seek = function(channel, control, value, status, group) {
+DDJ200.eq = function(channel, control, value, status, group) {
+    var val = script.absoluteNonLin(value, 0, 1, 4);
+    var eq = (control === 0x0B) ? 2 : 1;
+    if (control === 0x07) eq = 3;
```
```suggestion
    if (control === 0x07) {
        eq = 3;
    }
```

Holzhaus commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -46,14 +140,220 @@ DDJ200.seek = function(channel, control, value, status, group) {
+        var vgroup = "[Channel" + vDeckNo + "]";
+        var pfl = ! engine.getValue(vgroup, "pfl");
+        engine.setValue(vgroup, "pfl", pfl);
+        if (! DDJ200.fourDeckMode) {
```
```suggestion
        if (!DDJ200.fourDeckMode) {
```

Holzhaus commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -46,14 +140,220 @@ DDJ200.seek = function(channel, control, value, status, group) {
+                      (engine.getValue(vgroup, "cue_point") !== -1));
+    midi.sendShortMsg(0x90 + d, 0x58, 0x7F * engine.getValue(vgroup,
+        "sync_enabled"));
+    if (! DDJ200.fourDeckMode) {
```
```suggestion
    if (!DDJ200.fourDeckMode) {
```

Holzhaus commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -46,14 +140,220 @@ DDJ200.seek = function(channel, control, value, status, group) {
+        var vgroup = "[Channel" + vDeckNo + "]";
+        var playing = engine.getValue(vgroup, "play");
+        engine.setValue(vgroup, "play", ! playing);
+        if (engine.getValue(vgroup, "play") === playing)
+            engine.setValue(vgroup, "play", ! playing);
```
```suggestion
        if (engine.getValue(vgroup, "play") === playing) {
            engine.setValue(vgroup, "play", ! playing);
        }
```

Holzhaus commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -46,14 +140,220 @@ DDJ200.seek = function(channel, control, value, status, group) {
+        engine.setValue(vgroup, "play", ! playing);
+        if (engine.getValue(vgroup, "play") === playing)
+            engine.setValue(vgroup, "play", ! playing);
+        //midi.sendShortMsg(status, 0x0B, 0x7F * engine.getValue(vgroup, "play"));
```
```suggestion
```

fkbreitl committed:
Update res/controllers/Pioneer DDJ-200.midi.xml

Co-authored-by: Jan Holthuis <holthuis.jan@googlemail.com>

fkbreitl committed:
Update res/controllers/Pioneer-DDJ-200-scripts.js

Co-authored-by: Jan Holthuis <holthuis.jan@googlemail.com>

fkbreitl committed:
Update res/controllers/Pioneer-DDJ-200-scripts.js

Co-authored-by: Jan Holthuis <holthuis.jan@googlemail.com>

fkbreitl committed:
Update res/controllers/Pioneer-DDJ-200-scripts.js

Co-authored-by: Jan Holthuis <holthuis.jan@googlemail.com>

fkbreitl committed:
Update res/controllers/Pioneer-DDJ-200-scripts.js

Co-authored-by: Jan Holthuis <holthuis.jan@googlemail.com>

fkbreitl committed:
Update res/controllers/Pioneer-DDJ-200-scripts.js

Co-authored-by: Jan Holthuis <holthuis.jan@googlemail.com>

Holzhaus commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -1,41 +1,135 @@
+    }, true);
+};
+
+DDJ200.shutdown = function() { DDJ200.LEDsOff(); };
```
```suggestion
DDJ200.shutdown = function() {
    DDJ200.LEDsOff();
};
```

Holzhaus reviewed (approved):
Just a minor comment, otherwise LGTM.

fkbreitl committed:
Update res/controllers/Pioneer-DDJ-200-scripts.js

Co-authored-by: Jan Holthuis <holthuis.jan@googlemail.com>

Holzhaus reviewed (approved):
Thank you, LGTM.

fkbreitl commented:
Thank you!
